crates/adapters/src/adhoc/table.rs

```diff
@@ -0,0 +1,295 @@
+        async fn send_batch(
+            tx: &Sender<datafusion::common::Result<RecordBatch>>,
+            projection: &Option<Vec<usize>>,
+            mut batch: RecordBatch,
+        ) -> datafusion::common::Result<()> {
+            // Apply projection if necessary, ideally we would be able to do this on the
+            // fly in the cursor, but that would require a lot of changes.
+            if let Some(keep_indices) = projection {
+                batch = batch.project(keep_indices).map_err(|e| {
+                    DataFusionError::Execution(format!("Unable to project record batch: {}", e))
+                })?;
+            }
+
+            tx.send(Ok(batch)).await.map_err(|e| {
+                DataFusionError::Execution(format!("Unable to send record batch: {}", e))
+            })?;
+
+            Ok(())
+        }
+
+        let mut builder = RecordBatchReceiverStreamBuilder::new(self.schema.clone(), 10);
+
+        // Returns a single batch when the returned stream is polled
+        let batch_reader = self.readers[partition].clone();
+        let schema = self.schema.clone();
+        let tx = builder.tx();
+        let projection = self.projection.clone();
+
+        let sas: SerdeArrowSchema = schema.fields().iter().as_slice().try_into().map_err(|e| {
+            DataFusionError::Internal(format!(
+                "Unable to construct SerdeArrowSchema for the provided schema: {}.",
+                e
+            ))
+        })?;
+
+        builder.spawn(async move {
+            const MAX_BATCH_SIZE: usize = 1 << 16;
+            let mut cursor = batch_reader
+                .cursor(RecordFormat::Parquet(sas.clone()))
+                .map_err(|e| DataFusionError::External(Box::new(e)))?;
+            let mut insert_builder = SendableArrowBuilder::new(sas)?;
+
+            let mut cur_batch_size = 0;
+            while cursor.key_valid() {
+                if !cursor.val_valid() {
+                    cursor.step_key();
+                    continue;
+                }
+                let mut w = cursor.weight();
+
+                // Skip deleted records.
+                if w < 0 {
+                    cursor.step_key();
+                    continue;
+                }
+
+                while w != 0 {
+                    cursor
+                        .serialize_key_to_arrow(&mut insert_builder.builder)
+                        .map_err(|e| {
+                            DataFusionError::Execution(format!(
+                                "Unable to serialize record to arrow: {}",
+                                e
+                            ))
+                        })?;
+                    cur_batch_size += 1;
+                    w -= 1;
+
+                    if cur_batch_size >= MAX_BATCH_SIZE {
+                        let batch = insert_builder.builder.to_record_batch().map_err(|e| {
+                            DataFusionError::Execution(format!(
```

Comment:
handles arbitrarily large weights. But we chunk batches into smaller ones since this is an async stream so we can avoid manifesting huge batches if e.g., we filter most results out later anyways (support for filter pushdown yet in our tables)